File: TopCoder/src/com/leetcode/tree/PathSum2.java
package com.leetcode.tree;

import java.util.ArrayList;
import java.util.List;
import java.util.Stack;

public class PathSum2 {

    public List<List<Integer>> pathSum(TreeNode root, int sum) {
        List<List<Integer>> retVal = new ArrayList<>();
        Stack<TreeNode> nodeStack = new Stack<>();
        Stack<Integer> sumStack = new Stack<>();
        Stack<List<Integer>> pathStack = new Stack<>();
        nodeStack.push(root);
        sumStack.push(root.val);
        List<Integer> list = new ArrayList<>();
        list.add(root.val);
        pathStack.push(list);

        while (!nodeStack.isEmpty()) {
            TreeNode currentTop = nodeStack.pop();
            int currentSum = sumStack.pop();
            List<Integer> currentPath = pathStack.pop();
            if (currentTop.left == null && currentTop.right == null && currentSum == sum) {
                retVal.add(currentPath);
            }
            if (currentTop.left != null) {
                nodeStack.push(currentTop.left);
                sumStack.push(currentSum + currentTop.left.val);
                List<Integer> newPath = new ArrayList<>(currentPath);
                newPath.add(0, currentTop.left.val);
                pathStack.push(newPath);
            }

            if (currentTop.right != null) {
                nodeStack.push(currentTop.right);
                sumStack.push(currentSum + currentTop.right.val);
                List<Integer> newPath = new ArrayList<>(currentPath);
                newPath.add(0, currentTop.right.val);
                pathStack.push(newPath);
            }
        }
        return retVal;
    }





    public static void main(String[] args) {
        TreeNode root = new TreeNode(5);
        root.left = new TreeNode(4);
        root.right = new TreeNode(8);

        root.left.left = new TreeNode(11);
        root.right.right = new TreeNode(4);
        root.right.left = new TreeNode(13);

        root.left.left.left = new TreeNode(7);
        root.left.left.right = new TreeNode(2);

        root.right.right.left = new TreeNode(5);
        root.right.right.right = new TreeNode(1);
        List<List<Integer>> list = new PathSum2().pathSum(root, 22);
        System.out.println(list);
    }

}

File: TopCoder/python/ntree/PathSum2.py
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution:
    def pathSum(self, root: TreeNode, sum):
        results = []
        result = []
        self.helper1(root, [], results, 0, sum)
        return results

    def helper(self, root, result, results, currentSum, sum):
        if not root:
            return
        if not root.left and not root.right:
            if currentSum == sum:
                results.append(result)
            return

        if root.left:
            self.helper(root.left, result + [root.left.val], results, currentSum + root.left.val, sum)
        if root.right:
            self.helper(root.right, result + [root.right.val], results, currentSum + root.right.val, sum)

    def helper1(self, root, so_far, results, currentSum, sum):
        if not root:
            return
        if not root.left and not root.right and sum == root.val:
            so_far.append(root.val)
            results.append(so_far[:])
            so_far.pop()
            return
        so_far.append(root.val)
        if root.left:
            self.helper1(root.left, so_far, results, currentSum , sum - root.val)
        if root.right:
            self.helper1(root.right, so_far, results, currentSum, sum - root.val)
        so_far.pop()

if __name__ == '__main__':
    root = TreeNode(5)
    root.left = TreeNode(4)
    root.right = TreeNode(8)

    root.left.left = TreeNode(11)
    root.left.left.left = TreeNode(7)
    root.left.left.right = TreeNode(2)

    root.right.left = TreeNode(13)
    root.right.right = TreeNode(4)
    root.right.right.left = TreeNode(5)
    root.right.right.right = TreeNode(1)

    sol = Solution()
    res = sol.pathSum(root, 22)
    print(res)
